[WIP] New package: godot-beta

dataCobra opened the pull request:
#### Testing the changes
- I tested the changes in this PR: **briefly**

#### New package
- This new package conforms to the [package requirements](https://github.com/void-linux/void-packages/blob/master/CONTRIBUTING.md#package-requirements): **YES**

<!--
#### Local build testing
- I built this PR locally for my native architecture, (x86_64-glibc)

To check which changes to the template need to be made for Godot 4 to work.
I created the new package godot-beta which can exist beside the regular package.

First and foremost, this PR is about testing. Since the package only can be build on native architectures. We could consider adding this PR to Void but since beta will be obsolete with Godot 4 release I wouldn't recommend it.

Please test the PR on different architectures and give feedback here about the usability.

[ci skip]

classabbyamp commented:
> We could consider adding this PR to Void but since beta will be obsolete with Godot 4 release I wouldn't recommend it.

betas wouldn't be accepted anyways

dataCobra commented:
Everyone who is willing to test the building and try out the PR please let me know if everything works as expected.

dataCobra commented:
For me it works on x86_64-glibc

dataCobra commented:
For some currently unknown reason the new version of embree is no longer working with Godot.

Because of this I've changed the configuration so it fallsback to builtin embree for now.

jstnas commented:
Tested this on `x86_64-musl` and it does work, but I had to add `libexecinfo-devel` to the `makedepends`

dataCobra commented:
Hey @jstnas,

thank you for testing the PR.

Can you please test again?

I've now added `libexecinfo-devel` for all __*-musl__ architectures.

jstnas commented:
I've rebuilt the package and everything's working as expected.

zen0bit commented:
4.01 is out

jstnas commented:
testing 4.0.1, Godot crashed when built without `speech-dispatcher-devel`

eljamm commented:
I tested 4.0.1 on x86_64-glibc and everything seems to be working fine.

dataCobra commented:
I've pushed the new template for 4.0.1 taking the dependency for musl into account.

Can you please test the package again?

dataCobra commented:
I forgot to change the checksum...

Will push again. ^^

dataCobra commented:
Alright, now with the correct checksum. :+1:

dataCobra commented:
I've tested at least building successful for x86_64, i686, x86_64-musl. :slightly_smiling_face:

fgietzen commented:
Seems to work for me as well on x86_64. I was able to open and run a small test project without any problems.

zen0bit commented:
Tested on x64 glibc

just update PR name to be same as commit message..
 And LGTM..

Newchair2644 commented:
Just tested for x64 glibc on sway and I am getting this error when godot crashes trying to launch a project from the project selector.

``` err
Godot Engine v4.0.1.stable.custom_build - https://godotengine.org
xkbcommon: ERROR: couldn't find a Compose file for locale "en_US.utf-8" (mapped to "en_US.utf-8")
libspeechd.so.2: cannot open shared object file: No such file or directory
OpenGL API 3.3 (Core Profile) Mesa 22.3.5 - Compatibility - Using Device: Intel - Mesa Intel(R) HD Graphics 3000 (SNB GT2)
 
Editing project: /home/newchair/Code/games/hi
Godot Engine v4.0.1.stable.custom_build - https://godotengine.org
xkbcommon: ERROR: couldn't find a Compose file for locale "en_US.utf-8" (mapped to "en_US.utf-8")
libspeechd.so.2: cannot open shared object file: No such file or directory
ERROR: Required extension VK_KHR_surface not found, is a driver installed?
   at: _initialize_instance_extensions (drivers/vulkan/vulkan_context.cpp:468)
ERROR: Could not initialize Vulkan
   at: DisplayServerX11 (platform/linuxbsd/x11/display_server_x11.cpp:5465)
Your video card drivers seem not to support the required Vulkan version.

If possible, consider updating your video card drivers or using the OpenGL 3 driver.

You can enable the OpenGL 3 driver by starting the engine from the
command line with the command:

    "godot" --rendering-driver opengl3

If you recently updated your video card drivers, try rebooting.
ERROR: Unable to create DisplayServer, all display drivers failed.
   at: setup2 (main/main.cpp:2017)
ERROR: Pages in use exist at exit in PagedAllocator: N7Variant5Pools11BucketLargeE
   at: ~PagedAllocator (./core/templates/paged_allocator.h:140)
ERROR: Pages in use exist at exit in PagedAllocator: N7Variant5Pools12BucketMediumE
   at: ~PagedAllocator (./core/templates/paged_allocator.h:140)
ERROR: Pages in use exist at exit in PagedAllocator: N7Variant5Pools11BucketSmallE
   at: ~PagedAllocator (./core/templates/paged_allocator.h:140)
ERROR: BUG: Unreferenced static string to 0: interface_added
   at: unref (core/string/string_name.cpp:131)
   ```

Newchair2644 commented:
Getting the same error when testing the 4.0 binary on the godot website so its not an issue with this package.

eljamm commented:
This error seems to be related to your system's locale

    xkbcommon: ERROR: couldn't find a Compose file for locale "en_US.utf-8" (mapped to "en_US.utf-8")

Maybe your should set `LANG=en_US.UTF-8` in `/etc/locale.conf`[^1], and see if that solves it.

[^1]: https://docs.voidlinux.org/config/locales.html#setting-the-system-locale

---

Perhaps `speech-dispatcher` should be explicitly added as a runtime dependency. If I uninstall it from my system I also get this message:

    libspeechd.so.2: cannot open shared object file: No such file or directory

---

Either your graphics card is old and does not support Vulkan or your drivers are outdated. If your GPU does support Vulkan then try updating your drivers.

    ERROR: Required extension VK_KHR_surface not found, is a driver installed?
    at: _initialize_instance_extensions (drivers/vulkan/vulkan_context.cpp:468)
    ERROR: Could not initialize Vulkan
    at: DisplayServerX11 (platform/linuxbsd/x11/display_server_x11.cpp:5465)
    Your video card drivers seem not to support the required Vulkan version.

Newchair2644 commented:
You are 100% percent correct, my poor old intel hd 3000 igpu does not support vulkan. I apologize for adding irrelevant information to this PR. Thanks!

eljamm commented:
@Newchair2644 Your contribution brought up a dependency issue and provided insight on old hardware, so for me it's very much appreciated.

There is currently an open issue for this problem, upstream, if you want to follow: https://github.com/godotengine/godot/issues/58927

It's also worth noting that not all features are supported yet by the OpenGL 3 renderer[^1], as they're being worked on for version 4.1[^2], so you might want to stick with Godot 3 for now.

---

On that note, perhaps we should add an install message, like:

    If your graphics card does not support Vulkan, you need to use
    the OpenGL 3 renderer:
    
    $ godot --rendering-driver opengl3
    
    More details: https://github.com/godotengine/godot/issues/58927

[^1]: https://godotengine.org/article/status-of-opengl-renderer/
[^2]: https://github.com/orgs/godotengine/projects/33/views/1

dataCobra commented:
Hey @eljamm,

should I add `speech-dispatcher` as a default runtime dependency for all architectures?

Do you know how I add an install message to the package? I haven't done this before.

dataCobra commented:
I updated the title and the package to godot 4.0.2. :+1:

eljamm commented:
@dataCobra

Since it's not automatically detected, I think it should manually be added as a runtime dependency.
It's used by godot for TTS[^1], so not having it installed would leave this feature missing.

For the install message, you just have to put it inside an `INSTALL.msg` file[^2].

[^1]: https://docs.godotengine.org/en/stable/tutorials/audio/text_to_speech.html#requirements-for-functionality
[^2]: https://github.com/void-linux/void-packages/blob/master/Manual.md#installmsg-and-removemsg-files

dataCobra commented:
Hey everyone,

I've changed all the necessary stuff and added speech-dispatcher as a dependency.

I think we're good to go. Could others please test the PR one more time so we can confirm a working package?

dataCobra commented:
Hello everyone,

I've updated the package to version 4.0.3 of godot game engine.

bilebucket commented:
I had a brief test and the package indeed seems to work just fine.

eljamm commented:
Everything seems good to me as well.

obunden commented:
Would you guys say that this is ready for merging? Or is something untested?

bilebucket commented:
I'd say that it's good to merge. I've worked on a 2D project of mine with this package for some 7 hours, no instability or crashes occurred. I also briefly tested some of the available 3D demos and they seem to work well too. @dataCobra thank you for maintaining the Godot package!

leahneukirchen commented on srcpkgs/godot/template:
```diff
@@ -1,40 +1,41 @@
 LDFLAGS+=" -pie"
 
+case "$XBPS_TARGET_MACHINE" in
+	*-musl) makedepends+=" libexecinfo-devel speech-dispatcher-devel";;
```
Why is speech-dispatcher-devel only needed on musl?

Why is libexecinfo needed when the crash handler is disabled below?

dataCobra replied:
That is a good question. I'll try building without `libexecinfo-devel` and add `speech-dispatcher-devel` to the make dependencies.

dataCobra replied:
I don't know why the crash handler was disabled in the first place. Older commits don't explain this change.

If it's okay for you I would remove the crash handler change and move speech-dispatcher-devel and libexecinfo-devel to make_depends.

Note: I tested building with musl with the crash handler change still in place. musl does indeed need libexecinfo-devel and glibc doesn't need it for unknown reasons.

dataCobra replied:
I've already tested building without the crash handler change successfully for musl and glibc.

leahneukirchen replied:
> glibc doesn't need it for unknown reasons.

Well glibc implements this internally.

dataCobra replied:
I've now changed the template and did the following:

1. Get rid of the crash handler change.
2. Only `libexecinfo-devel` will now only be used for musl builds. (Only needed for musl)
3. `speech-dispatcher-devel` is now a dependency for all architectures.

I've successfully build natively x86, x86_64 and x86_64-musl with these changes.

leahneukirchen commented:
Please rebase and push so we can run CI for once. (I'm not 100% sure libexecinfo is picked up.)

dataCobra commented:
> Please rebase and push so we can run CI for once. (I'm not 100% sure libexecinfo is picked up.)

CI is now running. :)

I hope we get a good outcome.

astralchan commented:
> Well i686 fails but I don't quite understand what the problem is.
> 
> The system should be building x86 and not x86_64 but compliance about x86_64 not build?
> 
> As said before, building here natively x86 did work.

https://docs.godotengine.org/en/stable/contributing/development/compiling/introduction_to_the_buildsystem.html#architecture

> # Architecture
> The `arch` option is meant to control the CPU or OS version intended to run the binaries. It is focused mostly on desktop platforms and ignored everywhere else.
>
> Supported values for the arch option are **auto**, **x86_32**, **x86_64**, **arm32**, **arm64**, **rv64**, **ppc32**, **ppc64** and **wasm32**.
>
> ```
> scons platform=<platform> arch={auto|x86_32|x86_64|arm32|arm64|rv64|ppc32|ppc64|wasm32}
> ```
>
> This flag appends the value of `arch` to resulting binaries when relevant. The default value `arch=auto` detects the architecture that matches the host platform.

I don't know if i686 is a supported built target.

astralchan commented on srcpkgs/godot/template:
```diff
@@ -1,41 +1,38 @@
-	vsed -e 's/#ifdef CRASH_HANDLER_ENABLED/#if defined(CRASH_HANDLER_ENABLED) \&\& defined(__GLIBC__)/' \
-		-i platform/x11/crash_handler_x11.cpp
-}
+case "$XBPS_TARGET_MACHINE" in
```
Since this is targeting the libc, specifically, wouldn't `XBPS_TARGET_LIBC` be more accurate? Something like:

```sh
if [ "$XBPS_TARGET_LIBC" = "musl" ]; then
    makedepends +=" libexecinfo-devel"
fi
```

dataCobra replied:
Thank you for the help.

I've changed this in the new commit.

astralchan commented on srcpkgs/godot/template:
```diff
@@ -1,41 +1,38 @@
+case "$XBPS_TARGET_MACHINE" in
+	*-musl) makedepends+=" libexecinfo-devel ";;
+esac
 
```
For targeting i686, I might try:

```sh
if [ "$XBPS_TARGET_WORDSIZE" = "32" ]; then
    make_build_args+=" arch=x86_32"
fi
```

astralchan replied:
Covering all cases could look like:

```sh
case "$XBPS_TARGET_MACHINE" in
    x86) arch="x86_32" ;;
    x86_64*) arch="x86_64" ;;
    armv7*) arch="arm64" ;;
    ppc*) arch="ppc64" ;;
   #.
   #.
   #.
   # etc
esac

make_build_args+=" arch=$arch"
```

dataCobra replied:
I've decided to use the second method to make sure everything is covered correct.

leahneukirchen commented on srcpkgs/godot/template:
```diff
@@ -46,9 +43,10 @@ do_install() {
-		x86_64*|aarch64*) vbin bin/godot.x11.opt.tools.64 godot;;
-		ppc64*) vbin bin/godot.x11.opt.tools.ppc64 godot;;
-		*) vbin bin/godot.x11.opt.tools.32 godot;;
+		x86_64*|aarch64*) vbin bin/godot.linuxbsd.editor.x86_64 godot;;
```
This sounds wrong to me...

dataCobra replied:
What do you mean with this sounds wrong?

I don't see an issue here?

leahneukirchen replied:
Why would the arm64 binary be called x86_64?

dataCobra replied:
Oh now I see what you mean...

Yeah well that is indeed a problem.

I'll change it. Thank you for bringing that to my attention. :)

dataCobra replied:
Fixed in the new commit.

dataCobra commented:
It seems like with all the great help we've a functioning update.

If you stumble upon any other issues please let me know. Otherwise I would say we're finally ready for merge. :slightly_smiling_face:

astralchan commented on srcpkgs/godot/template:
```diff
@@ -1,41 +1,49 @@
 distfiles="https://github.com/godotengine/godot/archive/${version}-stable.tar.gz"
-checksum=164523c1c8aef0b69b135645794f5bece3f63788556a56aa293c118cde457023
+checksum=8d629aa4910bd0b13354a3b96c253fb62df130ff8acb510c81e23cd100bfdc9f
 nocross=https://build.voidlinux.org/builders/armv7l_builder/builds/6342/steps/shell_3/logs/stdio
```
If `nocross` is not empty, why even bother trying to set things for cross build?
